lib/membrane/rtp/vad.ex

```diff
@@ -132,85 +101,37 @@ defmodule Membrane.RTP.VAD do
 
   defp handle_vad(buffer, rtp_timestamp, level, state) do
     state = %{state | current_timestamp: rtp_timestamp}
-    state = filter_old_audio_levels(state)
     state = add_new_audio_level(state, level)
-    audio_levels_vad = get_audio_levels_vad(state)
+
+    {trimmed_queue, audio_levels_vad} =
+      IsSpeakingEstimator.trim_queue_and_estimate_vad(
+        state.audio_levels,
+        state.vad_threshold + 127
+      )
+
+    state = update_queue(trimmed_queue, state)
     actions = [buffer: {:output, buffer}] ++ maybe_send_event(audio_levels_vad, state)
     state = update_vad_state(audio_levels_vad, state)
     {actions, state}
   end
 
-  defp filter_old_audio_levels(state) do
-    Enum.reduce_while(state.audio_levels, state, fn {level, timestamp}, state ->
-      if Ratio.sub(state.current_timestamp, timestamp)
-         |> Ratio.gt?(state.rtp_timestamp_increment) do
-        {_level, audio_levels} = Qex.pop(state.audio_levels)
-
-        state = %{
-          state
-          | audio_levels_sum: state.audio_levels_sum - level,
-            audio_levels_count: state.audio_levels_count - 1,
-            audio_levels: audio_levels
-        }
-
-        {:cont, state}
-      else
-        {:halt, state}
-      end
-    end)
-  end
-
   defp add_new_audio_level(state, level) do
-    audio_levels = Qex.push(state.audio_levels, {-level, state.current_timestamp})
+    audio_levels = Qex.push_front(state.audio_levels, {-level, state.current_timestamp})
 
-    %{
-      state
-      | audio_levels: audio_levels,
-        audio_levels_sum: state.audio_levels_sum + -level,
-        audio_levels_count: state.audio_levels_count + 1
-    }
-  end
-
-  defp get_audio_levels_vad(state) do
-    if state.audio_levels_count >= state.min_packet_num and avg(state) >= state.vad_threshold,
-      do: :speech,
-      else: :silence
+    %{state | audio_levels: audio_levels}
   end
 
-  defp avg(state), do: state.audio_levels_sum / state.audio_levels_count
+  defp update_queue(new_queue, state), do: %{state | audio_levels: new_queue}
```

Comment:
This helper is a bit of an overkill, plus it's not used everywhere you update the queue, I'd got rid of it